packages/salesforcedx-vscode-apex-replay-debugger/src/telemetry/telemetry.ts

```diff
@@ -0,0 +1,62 @@
+import vscode = require('vscode');
+import TelemetryReporter from 'vscode-extension-telemetry';
+
+export class TelemetryService {
+  private static instance: TelemetryService;
+  private context: vscode.ExtensionContext | undefined;
+  private reporter: TelemetryReporter | undefined;
+  private isTelemetryEnabled: boolean;
+
+  constructor() {
+    this.isTelemetryEnabled = false;
+  }
+
+  public static getInstance() {
+    if (!TelemetryService.instance) {
+      TelemetryService.instance = new TelemetryService();
+    }
+    return TelemetryService.instance;
+  }
+
+  public initializeService(
```

Comment:
I see some opportunities to reduce code duplication. I understand why this class needs to be in the context of each extension (so that it reads the right package.json using require, which is static).

But for the other methods like `sendCommandEvent`, `sendExtensionActivationEvent`, and `sendExtensionDeactivationEvent`, we might be able to either:

1. Put them into a base class that this class extends from (see https://github.com/forcedotcom/salesforcedx-vscode/blob/develop/packages/salesforcedx-vscode-lwc-next/src/commands/forceLightningLwcCreate.ts#L72)
1. Use constructor dependency injection and inject different loggers into this class in the initializeService. Basically, each logging method that you have defined so far only needs access to `this.reporter` and `this.isTelemetryEnabled` so those can be passed in.

Which do you prefer?

Reply:
Agreed, I'll try extending from a class since it seems easier to read.

Reply:
I'm going with the constructor dependency injection approach since I failed miserably on setting the core extension as a test dependency for the base class approach.